portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -60,6 +63,14 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String message = getParameter(request, "text-input", "");
+
+    // Sentiment Analysis
+    Document doc = Document.newBuilder().setContent(message).setType(Document.Type.PLAIN_TEXT).build();
+    LanguageServiceClient languageService = LanguageServiceClient.create();
+    Sentiment sentiment = languageService.analyzeSentiment(doc).getDocumentSentiment();
+    float score = sentiment.getScore();
+    languageService.close();
+
```

Comment:
I was going to comment that I don't see a print statement as indicated in the PR description, which was initially confusing. I looked at the commit history and noticed that you took it out in the last commit in preparation for the next PR. I probably would have chosen to leave the printing in for this PR for a few reasons:

- consistency with PR description (you could also have changed the PR description to match the state of the PR of course, but see the next comment)
- without the print, there's no way to test that the code actually works, which can make reviewers nervous :). It's always a good practice to include tests (or at the very least, testability) in every PR.
- from a code health point of view, `score` is an unused variable, which many tools will complain about (we wouldn't actually be able to submit this code at Google for that reason)

I'll very frequently include some form of testing results (screenshot, link to some artifact that shows the code works, etc.) as evidence of the quality/completeness in my change descriptions.

Having said all that, I understand what you've done, and this looks fine!

A question to get you thinking...why have you chosen to perform the sentiment in the post handler instead of say the get handler? (I'm not asking you to change anything, just to articulate the reasons for the decision you've made --- or if the decision was not completely conscious, to think about it and see if it's a supportable decision --- so that we can discuss design principles a bit.

Reply:
Yes, I am not sure why I deleted the print statement... That was supposed to be for this PR.
Also, I am confused specifically about the testing results you mentioned. From my understanding, I should take a screenshot of the output where it prints the score in the terminal and attach it to my PR. Is this correct?
Hmm, good question. That was something I thought about a bit, because I personally would have done it in the get, but in the example I saw it in the "post". I'll first talk about why I think it would be better to keep it in the "get". I thought it would be a waste of resources to store it in the Datastore and it would be simpler (and maybe less code) to just compute the score when you call the "get". However, I think the example did it in the "post" because maybe otherwise there would be some delay for the user when the score comes to the screen. If I were to do it in "get", it may take more time to calculate the sentiment score for potentially thousands (at a large scale) of comments. This would not be the case in the "post" because we are only calculating the sentiment score for one comment at that specific moment.

Reply:
Re: testing....it's always a good idea to demonstrate that the code you're submitting works, by showing reviewers the observable results of the code. In the past, this is why I would ask for you to sync your app to the version of code I was running. In this case, the "observable" of the code would have been the print statement and a screenshot would have been appropriate.

Re: get vs post for sentiment analysis. I think you've thought quite clearly about the issues, nice! Let me respond a bit.

Your statement that it would be "waste of resources" to put the sentiment in datastore is completely accurate, *for some values of "resources"* :). In fact, there are lots of different resources that need to be considered:

- storage (space) - this can either be memory or persistent storage, and is the one you were focusing on in your initial statement
- time - this can be CPU time or "wall clock" time - you noted this when you discussed delay. When writing a UI-based application, this is very often a critical factor, because it dramatically affects how responsive the UI is perceived to be, which is a very important predictor of user satisfaction

And of course, everything translates to money :). Storage costs money, CPU costs money, having happy users makes money - but they have different costs. In a cloud environment, each API call (Datastore, sentiment analysis, maps) has its own billing profile. It could be an interesting exercise to research the billing for these different APIs and see if this is a compelling argument. In a non-cloud environment, this translates to server and power costs.

From an architectural standpoint, your realization that this is a "write once, read many" system is key. Assuming that you have lots of visitors to your site, having to call the sentiment API N times on GET vs only 1 for POST to get the same answer could get very expensive.

Reply:
Ah, thank you for the detailed explanation. It makes a lot of sense now.